pkg/store/proxy_heap.go

```diff
@@ -27,104 +26,93 @@ import (
 	"github.com/thanos-io/thanos/pkg/tracing"
 )
 
-// dedupResponseHeap is a wrapper around ProxyResponseHeap
-// that removes duplicated identical chunks identified by the same labelset and checksum.
-// It uses a hashing function to do that.
 type dedupResponseHeap struct {
 	h *ProxyResponseHeap
 
-	responses []*storepb.SeriesResponse
+	bufferedSameSeries []*storepb.SeriesResponse
 
-	previousResponse *storepb.SeriesResponse
-	previousNext     bool
+	bufferedResp []*storepb.SeriesResponse
+	buffRespI    int
+
+	prev *storepb.SeriesResponse
+	ok   bool
 }
 
+// NewDedupResponseHeap returns a wrapper around ProxyResponseHeap that merged duplicated series messages into one.
+// It also deduplicates identical chunks identified by the same checksum from each series message.
 func NewDedupResponseHeap(h *ProxyResponseHeap) *dedupResponseHeap {
```

Comment:
Why does this code need to change? From what I understand we just need to fix the dedup iterator so that it splits overlapping chunks into separate streams. I don't understand what the changed version of dedup heap does now.

Reply:
1. There was a bug in aggregating the series into one series with non-series responses in between.
2. Two big complexity points (`Next` and `At`), instead of focusing complexity on one place (Next) and keeping At trivial (less cognitive load).
3. I found previous code less readable than my version (too ambiguous state, too many if else, complex deferred logic), but I guess it's a biased opinion. 🙈